### PR TITLE
Add tracers to zk server

### PR DESCRIPTION
**Pull request by gianbelinche**
## What ❔
This PR adds tracers for the zk server, it is basically a merge between branches `era_vm_zk_server_v2` and `implement_tracers` with a few changes to make it work.

All tests under `zk test i server` pass.

Note: A couple of tracers, `StorageInvocations` and `ValidationTracer`, were not needed to make te tests pass, so they were not implemented for era vm, `TODO` comments were left for them in case we need them in the future.

The timeout for tests was increased in order to make the tests pass, however this is due to the era vm being too slow, this could be changed later once it is well optimized.
 
Related to:
https://github.com/lambdaclass/era-compiler-tester/pull/24
https://github.com/lambdaclass/era_vm/pull/221

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/lib/multivm/src/vm_instance.rs [hunk 1]
```diff
@@ -263,8 +265,8 @@ impl<S: ReadStorage, H: HistoryMode> VmInstance<S, H> {
             VmVersion::Vm1_5_0IncreasedBootloaderMemory => match mode {
                 FastVmMode::Old => Self::new(l1_batch_env, system_env, storage_view),
                 FastVmMode::New => {
```
Comment (MarcosNicolau) [inline core/lib/multivm/src/vm_instance.rs:267]: Could we add a new mode for our vm instead of removing the `vm_fast`? If it is too complicated we can leave it for another pr though.

Comment (gianbelinche) [reply core/lib/multivm/src/vm_instance.rs:267]: It should not be difficult, I'll add it

Comment (gianbelinche) [reply core/lib/multivm/src/vm_instance.rs:267]: At the end it was not as easy as intended, there a two options for this, one is adding a new value under `core/lib/basic_types/src/vm.rs` `FastVmMode` , however we are not actually a new version of the fast vm, so I'm not sure this is the correct approach, also this would also mean a change in the `zksync-protobuf-config` repository.

Another option is to add our vm under `core/lib/basic_types/src/vm.rs` `VmVersion`, but that struct is directly related to `core/lib/basic_types/src/protocol_version.rs` `ProtocolVersionId`, and it does not seem right to add a new protocol version for our vm.

**Review by MarcosNicolau — APPROVED** [review]

File: core/lib/multivm/src/vm_instance.rs (hunk 1 above)
Comment (MarcosNicolau) [reply core/lib/multivm/src/vm_instance.rs:267]: Yeah, let's not worry about that for now.

File: core/lib/multivm/src/glue/tracers/mod.rs [hunk 2]
```diff
@@ -57,6 +58,10 @@ pub trait IntoLatestTracer<S: WriteStorage, H: HistoryMode> {
     fn latest(&self) -> crate::vm_latest::TracerPointer<S, H::Vm1_5_0>;
 }
 
+pub trait IntoEraVmTracer<S: WriteStorage, H: HistoryMode> {
+    fn era_vm(&self) -> Box<dyn crate::era_vm::tracers::traits::VmTracer<S>>;
```
Comment (Oppen) [inline core/lib/multivm/src/glue/tracers/mod.rs:62]: I find the method name a bit confusing. Maybe `into_era_vm`?

Comment (Oppen) [reply core/lib/multivm/src/glue/tracers/mod.rs:62]: Also, maybe implementing `From` is enough? I'm not 100% sure, just a suggestion.

Comment (Oppen) [reply core/lib/multivm/src/glue/tracers/mod.rs:62]: If it's OK with you, I may check the branch out and try it locally.

Comment (gianbelinche) [reply core/lib/multivm/src/glue/tracers/mod.rs:62]: That is the way they do it for the other vms, I think it is better to keep the same type of implementation, that doesn't mean your suggestion won't work though.

File: core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                    .last()
+                    .map(|call| call.frame.gas_left.0.saturating_add(current_ergs))
+                    .unwrap_or(current_ergs) as u64;
+
+                // we need to to this cast because `Call` uses another library
+                let far_call_variant = match far_call as u8 {
+                    0 => FarCallOpcode::Normal,
+                    1 => FarCallOpcode::Delegate,
+                    2 => FarCallOpcode::Mimic,
+                    _ => unreachable!(),
```
Comment (fkrause98) [inline core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs:44]: We can avoid the unreachable! macro here, let's use the variant defined here on zksync-era to match with ours, even if we have to use total namespaces.

File: core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        // Actually it's a call of the constructor.
+        // And at this stage caller is user and callee is deployed contract.
+        let call_type = if let CallType::Call(far_call) = current_call.r#type {
+            if matches!(far_call, FarCallOpcode::Mimic) {
+                let previous_caller = execution
+                    .running_contexts
+                    .first()
+                    .map(|call| call.caller)
+                    // Actually it's safe to just unwrap here, because we have at least one call in the stack
+                    // But i want to be sure that we will not have any problems in the future
```
Comment (fkrause98) [inline core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs:86]: Good job, unwrap_or is a good compromise here.

File: core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                if previous_caller == CONTRACT_DEPLOYER_ADDRESS {
+                    CallType::Create
+                } else {
+                    CallType::Call(far_call)
+                }
+            } else {
+                CallType::Call(far_call)
+            }
+        } else {
+            unreachable!()
```
Comment (fkrause98) [inline core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs:97]: If we don't do anything with another variant, we can replace this if match + unreachable with:` if let FarCallOpcode::Mimic = .... {}`

File: core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                        .unwrap()
+                        .read_unaligned_from_pointer(&fat_data_pointer)
+                        .unwrap(),
+                )
+            } else {
+                None
+            }
+        } else {
+            None
+        };
```
Comment (fkrause98) [inline core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs:146]: Consider this change, I think it's clearer:
```suggestion
              let output = match fat_data_pointer.is_pointer {
            true => {
                let fat_data_pointer = FatPointer::decode(fat_data_pointer.value);
                match (fat_data_pointer.len, fat_data_pointer.offset) {
                    (0, 0) => execution
                        .heaps
                        .get(fat_data_pointer.page)
                        .and_then(|ptr| ptr.read_unaligned_from_pointer(&fat_data_pointer).ok()),
                    _ => None,
                }
            }
            _ => None,
        };
```

File: core/lib/multivm/src/versions/era_vm/tracers/pubdata_tracer.rs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                let diff = StateDiffRecord {
+                    key,
+                    address,
+                    derived_key:
+                        zk_evm_1_5_0::aux_structures::LogQuery::derive_final_address_for_params(
+                            &address, &key,
+                        ),
+                    enumeration_index: vm
+                        .storage
+                        .borrow_mut()
```
Comment (fkrause98) [inline core/lib/multivm/src/versions/era_vm/tracers/pubdata_tracer.rs:78]: I don't like borrowing refcells outside of the implant block for this struct:
The reason is that this will panic in runtime if there is another living reference to it, which it could possibly happen given that this field is public, let's change this, do we need to change something else in our vm?

**Review by fkrause98 — CHANGES_REQUESTED** [review]

**Review by fkrause98 — APPROVED** [review]

File: core/lib/multivm/src/glue/tracers/mod.rs (hunk 2 above)
Comment (Oppen) [reply core/lib/multivm/src/glue/tracers/mod.rs]: Yes, if it's already done this way everywhere else it's good to be consistent.

File: core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                    CallType::Create
+                } else {
+                    CallType::Call(far_call)
+                }
+            } else {
+                CallType::Call(far_call)
+            }
+        } else {
+            return;
+        };
```
Comment (Oppen) [inline core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs:98]: You can reduce nesting a bit by using `let <destructuring pattern> else` here:
```suggestion
        let CallType::Call(far_call) = current_call.r#type else {
            return;
        }
        if matches!(far_call, FarCallOpcode::Mimic) {
            let previous_caller = execution
                .running_contexts
                .first()
                .map(|call| call.caller)
            // Actually it's safe to just unwrap here, because we have at least one call in the stack
            // But i want to be sure that we will not have any problems in the future
                .unwrap_or(current.caller);
            if previous_caller == CONTRACT_DEPLOYER_ADDRESS {
                CallType::Create
            } else {
                CallType::Call(far_call)
            }
        } else {
            CallType::Call(far_call)
        };
```

Comment (gianbelinche) [reply core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs:98]: The thing is that this code is exactly as is for the other vms, so I think it is better to leave it as is. The only change was the one from the unreachable, so we don't stop execution due to a tracer (even though in the other vms they still have it)
